# About This Project

Pipeline-prep project's goal is in two folds - first create a set reusable modules used to pre-index the reference 
before running a given data pipeline - second create an automation that triggers pre-indexing of the reference each time 
one of the following events happen: 
  1) A new version of the alignment tool was installed, or 
  2) A new release of the reference was downloaded

The goals :
  1) For each alignment tool, create a stable process that Pre-Indexes the reference genome/transcripome 
  2) Automate the Pre-indexing step
  ```
    a) Create a trigger to index the reference when a New Data version is detected
    b) Create a trigger to index the reference when a New Tool version is detected
  ```
  
Indexes are created by tool version -> data source version > organism-dataset. Each data source specifies
the format of each of its datasets and each tool has its own sets of command to index the reference data.

# Package Organization

The information under the root directory of the package is organized as follow:

 - [Package Main Config File](#package-main-config-file)(Configuration.cfg)
 - [New Data Release Index Trigger](#new-data-release-index-trigger)(newDatasetVersionTrigger.sh)  
 - [New Tool Version Index Trigger](#new-tool-version-index-trigger)(newToolVersionTrigger.sh)  
 - [Data Sources Base](#data-sources-base)(data_sources) 
 - [Indexers Base](#indexers-base)(tools)  

 


  See:
     [<img src="workflow.pdf">](workflow.pdf)

## Package Main Config File
## New Data Release Index Trigger
## New Tool Version Index Trigger
## Data Sources Base

Each data source is config file (tool.cfg) under data_sources/  - Each config file contains the following 
global environment variables:
  1) DATASETS - Mapping between each dataset and its file format 
  2) GENOME - AN ARRAY of the names of datasets that are genomic (the whole genome) - used as a flag for indexing options
  3) DATA_SOURCE - name of the data source
  4) ORGANISMS - list of organism as provived (datasets are stored by organisms)
  5) and more
  
### Data Sources Used To Test
  1) Ensembl
  2) Wormbase
  
## Indexers Base

Each tool is a directory under tools  and each directory contains the following files:

1)	A readme.md file
2)	A config file tool.cfg
3)	Refeence.csv file mapping tool to datasets - one reference.csv file for each data source
4)	The Index shell script to setup and index datasets

#### The Tool.cfg Config File
Each tool's config file contains global environment variables ranging from the tool local install path, datasets path,
tool indexes path,to index options.
  1) Tool Local Install config
  ```
     PACKAGE_INSTALL_BASE  - path to tool install directory
     TOOL_VERSION - name of the directory created by the indexer for this version of the tool
     INDEX_BASE  -- path to where indexes are stored
  ```
  2) Path To Datasets and logs
  ```
     FASTA_BASE -- path to datasets fasta files
     LOGS_BASE  -- path to logs generated by this tool
     
  ```
  3) Indexer config
  ```
     REFERENCE_FILE -- a list of reference.csv files indexed bu this tool - one for each data source
     TOOL_INDEX_EXEC -- the indexer (the executable)
     ALGO_GENOME  -- indexer option for genome indexing (bwa)
     ALGO_TRANSCRIPTOME -- indexer option for transcriptome indexing
     
  ```

##### Tool Install Setup



## Automate Pre-indexing step




