# Data Processing Pipelines

```
 "Pipelines are composed of discrete steps that can represent an algorithm, 
 a software tool, or a file format manipulation. These steps are applied to the primary data
 (generated from an experimental assay) to produce visualizable data" ~ENCODE
 ```
 ## Input
 Data from major experimental assay types generated by the project: 
 * RNA-seq, 
 * RAMPAGE1, 
 * ChIP-seq, 
 * DNase-seq, 
 * ATAC-seq2 ,
 * WGBS
 
 ## Type Of Pipelines
 
 * RNA-seq pipelines
 * RAMPAGE pipelines
 * Chromatin Immunoprecipitation pipelines 
 * DNA accessibility pipelines
 * DNA methylation pipeline
 
 ## Pipeline Steps
 * Pre-processing the input
 * Pre-Indexing reference genome/transcripome
 * Mapping reads to ref genome/transcriptome
 * features quantification
 * Features assembly
 * Features differenciation
 * Stats generation
 * Visualization
 
 ## Vesioning Pipelines
 
 ## RNA-Seq Pipelines
 * RNAs Long Reads - longer than 200 bp: for mRNAs (poly-A(+)), rRNA-depleted total RNA, or poly-A(-) RNA populations
     * Paired-end
     * Single-ended

 * RNAs Short reads - shorter than 200 bp: for mRNAs (poly-A(+)), rRNA-depleted total RNA, or poly-A(-) RNA populations
 * miRNA-seq: for microRNAs, around 22nt long, that are quantified from RNA-seq data
 * miRNA counts: a complement to miRNA-seq 

## RAMPAGE Pipelines

RAMPAGE (RNA Annotation and Mapping of Promoters for the Analysis of Gene Expression) is a very accurate sequencing approach to identify transcription start sites (TSSs) at base-pair resolution, to quantify their expression, and to characterize their transcripts. RAMPAGE uses direct cDNA evidence to link specific genes and their TSSs.

 ## USEFUL LINKS
 * Data Processing Pipelines: https://www.encodeproject.org/pipelines/
